### PR TITLE
Use unified cgroup hierarchy with nspawn --as-pid2

### PR DESCRIPTION
**Pull request by keszybz**

### CODE REVIEW

File: src/nspawn/nspawn.c [hunk 1]
```diff
@@ -412,15 +412,27 @@ static int custom_mount_check_all(void) {
...
+        const char *e, *var = "SYSTEMD_NSPAWN_UNIFIED_HIERARCHY";
         int r;
 
         /* Allow the user to control whether the unified hierarchy is used */
-        e = getenv("UNIFIED_CGROUP_HIERARCHY");
-        if (e) {
+
+        e = getenv(var);
+        if (!e) {
+                static bool warned = false;
```
Comment (filbranden) [inline src/nspawn/nspawn.c:422]: `warned` is set but never used... I'm thinking maybe you got this snippet from a loop? In any case, it's not really needed here...

File: src/nspawn/nspawn.c [hunk 2]
```diff
@@ -412,15 +412,27 @@ static int custom_mount_check_all(void) {
...
+                        log_info("$UNIFIED_CGROUP_HIERARCHY has been renamed to $SYSTEMD_NSPAWN_UNIFIED_HIERARCHY.");
+                        warned = true;
+                }
+        }
+
+        if (!isempty(e)) {
                 r = parse_boolean(e);
                 if (r < 0)
-                        return log_error_errno(r, "Failed to parse $UNIFIED_CGROUP_HIERARCHY.");
+                        return log_error_errno(r, "Failed to parse $%s: %m", var);
```
Comment (filbranden) [inline src/nspawn/nspawn.c:435]: I'm wondering if this should maybe always say `$SYSTEMD_NSPAWN_UNIFIED_HIERARCHY`, even if you're setting the legacy variable name? If you set the legacy one, you'll get both logs, so you'll figure out what it's referring to, no?

File: test/test-functions [hunk 3]
```diff
@@ -250,12 +250,12 @@ run_nspawn() {
     fi
 
     if [[ "$UNIFIED_CGROUP_HIERARCHY" = "hybrid" ]]; then
```
Comment (filbranden) [inline test/test-functions:252]: Should you support `$SYSTEMD_NSPAWN_UNIFIED_HIERARCHY` here as well? Maybe only `$SYSTEMD_NSPAWN_UNIFIED_HIERARCHY` if this is internal and we control it?

Right now it looks a bit confusing that we renamed it mostly everywhere, except for here...

File: src/nspawn/nspawn.c (hunk 2 above)
Comment (keszybz) [reply src/nspawn/nspawn.c:435]: It is possible to just get the error message, the other one has lower priority. So I think this message should be accurate. I really dislike it when tools give a message describing an option I didn't use.

File: test/test-functions (hunk 3 above)
Comment (keszybz) [reply test/test-functions:252]: UNIFIED_CGROUP_HIERARCHY=hybrid|default|unified|legacy is out internal test setting.
SYSTEMD_NSPAWN_UNIFIED_HIERARCHY=0|1 is the nspawn setting.

Actually, UNIFIED_CGROUP_HIERARCHY still appears in many many places.

File: src/nspawn/nspawn.c (hunk 1 above)
Comment (keszybz) [reply src/nspawn/nspawn.c:422]: It's a static variable so it get preserved between function calls. The goal is to emit the warning just once. The function is called twice, so the warning would appear twice without this check.

**Comment by keszybz** [comment]
@filbranden, @yuwata: another review would be appreciated.

File: src/nspawn/nspawn.c (hunk 1 above)
Comment (filbranden) [reply src/nspawn/nspawn.c:422]: Ah sorry I missed the `static` part at a glance... looks good.

File: test/test-functions (hunk 3 above)
Comment (filbranden) [reply test/test-functions:252]: I guess that looks good.

I think I was perceiving this as confusing, but it's actually splitting two separate things into two separate names, so I guess that's good.

**Pull request merged**